shim/shim.c

```diff
@@ -0,0 +1,292 @@
+	int sig;
+	char *buf;
+	int ret;
+	int cmd_type;
+	struct winsize ws;
+ 
+	while (read(pipe_fd[0], &sig, sizeof(sig)) != -1) {
+		if (sig == SIGWINCH ) {
+			cmd_type = WINSIZE;
+			if (ioctl(STDIN_FILENO, TIOCGWINSZ, &ws) == -1) {
+				fprintf(stderr, "Error getting the current window size: %s\n",
+					strerror(errno));
+				continue;
+			}
+			ret = asprintf(&buf, "{\"container_id\":\"%s\", \"row\":\"%d\", \"col\":\"%d\"}",
+                                                        container_id, ws.ws_row, ws.ws_col);
+		} else {
+			cmd_type = KILLCONTAINER;
+			ret = asprintf(&buf, "{\"container_id\":\"%s\", \"signal\":\"%d\"}",
+                                                        container_id, sig);
+		}
+		if (ret == -1) {
+			abort();
+		}
+
+		send_hyper_message(outfd, cmd_type, buf);
+		free(buf);
+        }
+}
+void
+handle_data(int infd, int outfd)
+{
+	ssize_t nread;
+	char buf[BUF_SIZE];
+	int ret;
+
+	while ((nread = read(infd, buf, BUF_SIZE))!= -1) {
+		ret = (int)write(outfd, buf, (size_t)nread);
+		if (ret == -1) {
+			fprintf(stderr, "Error writing to fd : %s\n", strerror(errno));
+			return;
+		}
+	}
+}
+
+int
+main(int argc, char **argv)
+{
+	int proxy_sockfd;
+	struct sockaddr_un proxy_address;
+	char* container_id;
+	char* proxy_sock_path;
+	struct pollfd poll_fds[MAX_POLL_FDS];
+	nfds_t nfds = 0;
+	int ret;
+	struct sigaction sa;
+
+	if (argc != 3) {
+		printf("Usage: cc-shim container_id proxy_socket_path");
+		exit(EXIT_FAILURE);
+	}
+
+	container_id = argv[1];
+	proxy_sock_path = argv[2];
+
+	/* Using self pipe trick to handle signals in the main loop, other strategy
+	 * would be to clock signals and use signalfd()/ to handle signals synchronously
+	 */
+	if (pipe(pipe_fd) == -1) {
+		printf("Error creating pipe\n");
+		exit(EXIT_FAILURE);
+	}
+
+	// Add read end of pipe to pollfd list and make it non-bocking
+	add_pollfd(poll_fds, &nfds, pipe_fd[0], POLLIN | POLLPRI);
+	set_fd_nonblocking(pipe_fd[0]);
+	set_fd_nonblocking(pipe_fd[1]);
+
+	sigemptyset(&sa.sa_mask);
+	sa.sa_flags = SA_RESTART;           /* Restart interrupted reads()s */
+	sa.sa_handler = signal_handler;
+
+	/* TODO: Add signal handler for all signals */
+	if (sigaction(SIGINT, &sa, NULL) == -1)
+		err_exit("sigaction");
+	if (sigaction(SIGTERM, &sa, NULL) == -1)
+		err_exit("sigaction");
+	if (sigaction(SIGWINCH, &sa, NULL) == -1)
+		err_exit("sigaction");
+
+	set_fd_nonblocking(STDIN_FILENO);
+	add_pollfd(poll_fds, &nfds, STDIN_FILENO, POLLIN | POLLPRI);
+
+#if 0
+	// Connect to the cc-proxy AF_UNIX server
+	proxy_sockfd = socket(AF_UNIX, SOCK_STREAM, 0);
+	strcpy(proxy_address.sun_path, proxy_sock_path);
+	//block waiting to connect to the cc_proxy
+	connect(proxy_sockfd, (struct sockaddr *)&proxy_address, sizeof(proxy_address));
+
+	// Add the proxy socket to the list of pollfds
+	add_pollfd(poll_fds, &nfds, proxy_sockfd, POLLIN | POLLPRI);
+#endif
+
+	/*	0 =>pipe_fd[0]
+		1 =>stdin
+		2 =>sockfd
+	*/
+
+	while (1) {
```

Comment:
The shim needs to shutdown when the container does (returning the containers exit code), but it doesn't look like that is currently handled?